block/submit.go

```diff
@@ -2,38 +2,95 @@ package block
 func (m *Manager) SubmitLoop(ctx context.Context) {
+	// ticker to limit the time between submissions
 	ticker := time.NewTicker(m.Conf.BatchSubmitMaxTime)
 	defer ticker.Stop()
 
-	// TODO: add submission trigger by batch size (should be signaled from the the block production)
+	// get produced size from the block production loop and signal to submit the batch when batch size reached
+	submitByAccumulatedSizeCh := make(chan bool, m.Conf.MaxSupportedBatchSkew)
+	go m.AccumulatedDataLoop(ctx, submitByAccumulatedSizeCh)
+
+	// defer func to clear the channels to release blocked goroutines on shutdown
+	defer func() {
+		for {
+			select {
+			case <-m.producedSizeCh:
+			case <-submitByAccumulatedSizeCh:
+			default:
+				return
+			}
+		}
+	}()
+
 	for {
 		select {
-		// Context canceled
 		case <-ctx.Done():
 			return
-		// trigger by time
-		case <-ticker.C:
-			err := m.HandleSubmissionTrigger(ctx)
-			if errors.Is(err, gerr.ErrAborted) {
-				continue
-			}
-			if errors.Is(err, gerr.ErrUnauthenticated) {
-				panic(fmt.Errorf("handle submission trigger: %w", err))
-			}
-			if err != nil {
-				m.logger.Error("handle submission trigger", "error", err)
+		case <-submitByAccumulatedSizeCh: // block production
+		case <-ticker.C: // max time
+			// reset the accumulated size when triggered by time,
+			// as we gonna submit all our data anyway
+			m.AccumulatedBatchSize.Store(0)
+		}
+
+		// modular submission methods have own retries mechanism.
+		// if error returned, we assume it's unrecoverable.
+		err := m.HandleSubmissionTrigger(ctx)
+		if err != nil {
+			panic(fmt.Errorf("handle submission trigger: %w", err))
+		}
+		ticker.Reset(m.Conf.BatchSubmitMaxTime)
+	}
+}
+
+// AccumulatedDataLoop is the main loop for accumulating the produced data size.
+// It is triggered by the ProducedSizeCh channel, which is triggered by the block production loop when a new block is produced.
+// It accumulates the size of the produced data and triggers the submission of the batch when the accumulated size is greater than the max size.
+// It also emits a health status event when the submission channel is full.
+func (m *Manager) AccumulatedDataLoop(ctx context.Context, toSubmit chan bool) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case size := <-m.producedSizeCh:
+			total := m.AccumulatedBatchSize.Add(size)
+
+			// Check if accumulated size is greater than the max size
+			// TODO: allow some tolerance for block size (e.g support for BlockBatchMaxSize +- 10%)
+			if total > m.Conf.BlockBatchMaxSizeBytes {
+				select {
+				case toSubmit <- true:
+					m.logger.Info("new batch accumulated, signal sent to submit the batch")
+				default:
+					m.logger.Error("new batch accumulated, but channel is full, stopping block production until the signal is consumed")
+					// emit unhealthy event for the node
+					evt := &events.DataHealthStatus{Error: fmt.Errorf("submission channel is full")}
```

Comment:
right.
I'll handle it in separate PR.

Reply:
It's actually not that simple.
As our process of submitting is "running forever", it might just happen that submission is slow, w/o any specific error.
If we want to use the `/health` endpoint for debugging, we should probably expose some metrics.
but I'm not sure it's the purpose of it